Better caching and other fixes

JuniorJPDJ opened the pull request:
- is_zipfile caching
- holding ZipExtFile per file handle, not per path (keeps file position even for the same file opened more than once) - IMHO it's better and more simple/clean solution than whole caching class
- PyCharm .gitignore
- requirements.txt
- typing fixes
- style fixes (PyCharm was literally crying about them)
- configurable cache size
- fixed fusepy package name (fuse is original upstream package name and fusepy is renamed version in ubuntu)
- block reads in non-seekable file (not gonna happen anyway)
- fixed race condition in access to the same ZipFile / ZipExtFile, which resulted in weird seeking, causing double ZipFile reads

Fixed #2, fixes #3, resolves last issues from #1

JuniorJPDJ committed:
Revert "Add second level of caching file objects"

This reverts commit 5df6ec3b

JuniorJPDJ committed:
PyCharm support and requirements for easier installation

JuniorJPDJ committed:
Better caching, saving open files and minor fixes

wvffle commented on ziprofs.py:
```diff
@@ -198,37 +167,43 @@ def readdir(self, path, fh):
         formatter_class=argparse.ArgumentDefaultsHelpFormatter)
-    parser.add_argument('ROOT', nargs='?', help="filesystem root")
-    parser.add_argument('MOUNTPOINT', nargs='?', help="filesystem mount point")
+    parser.add_argument('root', nargs='?', help="filesystem root")
+    parser.add_argument('mountpoint', nargs='?', help="filesystem mount point")
```
You've changed the argument names but you forgot to update the readme with the changed names

JuniorJPDJ committed:
Updated README.md (duck off @wvffle)

JuniorJPDJ committed:
Updated test to newer ubuntu

wvffle commented:
:D

JuniorJPDJ commented:
```
$ ./runtest.sh 
Running ziprofs tests...
Mounting filesystem
--------------------------
Running test: zip is directory
PASS
--------------------------
Running test: tree
PASS
--------------------------
Running test: reading file content #1
PASS
--------------------------
Running test: reading file content #2
PASS
--------------------------
Running test: running script
PASS
Killing ziprofs
5/5 tests passed.
```
Funny, tests passed on my machine..

JuniorJPDJ committed:
Handled ubuntu and upstream fusepy package name

JuniorJPDJ commented:
Ubuntu renamed fusepy's package from fuse to fusepy, arch have it like in upstream.

JuniorJPDJ committed:
Fix argument type

JuniorJPDJ committed:
Back-seeking unseekable file should be unsupported
Also fixed access call

JuniorJPDJ committed:
Do not allow simultaneous reads (fixes last problem from #1)

qu1ck commented on .gitignore:
```diff
@@ -2,3 +2,5 @@
 test/mnt/*
 objectdb
+venv/
+.idea/
```
Add a newline at the end.

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,84 @@
-
-    def infolist(self):
-        return self.zf.infolist()
+@lru_cache(maxsize=2048)
```
I don't think this is safe to cache directly without looking at file mtime. Underlying file system can change between calls to is_zipfile, what was a file can become a directory, a file can be in process of being written and if it wasn't a zipfile before can be one later etc.

Caching this is not saving anything for first call which is where the bulk of time will be anyway.

I think it's better to just leave this to kernel fs caching.

JuniorJPDJ replied:
When I do not use this, whole dir listing is noticeably slower.
I don't feel like leaving it to kernel fs caching is good solution.

I probably could get mtime in account and write own caching, but not atm.

Also.. I've weird idea..
What If we give mtime as argument to this function? It will cache only call with the same mtime.
What do you think?

qu1ck replied:
Looks like a good enough compromise.

JuniorJPDJ committed:
Use ZipFile's internal lock
Blocks simultaneous access to any file in ZipFile instance (#1)

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,84 @@
     def _add(self, path: str):
         if path in self.cache:
             return
-        with self.rwlock:
```
The lock here shouldn't be removed. In fact I don't have it locked down enough, there are still race conditions that can lead to incorrect behavior in get().
All access to cache dict should be guarded by lock, including reads. Otherwise cache entry can disappear between `if path in self.cache` and `if mtime > self.cache[path][0]` for example.

JuniorJPDJ replied:
Sorry, I just reverted this commit and forgot about locks.
Lock for cache will be back in the next commit :D

qu1ck commented on ziprofs.py:
```diff
@@ -120,23 +88,25 @@ def get_zip_path(path: str) -> str:
-        zip_path = self.get_zip_path(path)
-        if not zip_path:
+        if self.get_zip_path(path):
+            if mode & os.W_OK:
```
good catch

qu1ck commented on ziprofs.py:
```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
-        if not zip_path:
-            return os.open(path, flags)
-        return 0
+        if zip_path := self.get_zip_path(path):
```
No walruses please. They make code less readable in most cases.

qu1ck commented on ziprofs.py:
```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
+            fh = self._get_free_zip_fh()
+            zf = self.zip_factory.get(zip_path)
+            self._zip_file_fh[fh] = zf.open(path[len(zip_path) + 1:])
+            self._zip_lock_fh[fh] = Lock()
```
This should be guarded by a lock. File handle you generate can be reused between when you get it and then put it in cache.

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,84 @@
+        while True:
+            if i not in self._zip_file_fh and i not in self._zip_lock_fh:
+                return i
+            i += 2
```
This is shorter and easier to read:
```python
while i in self._zip_file_fh:
  i += 2
return i
```

JuniorJPDJ replied:
And that's how it looks now, you are talking about version few commits ago.
I removed whole dict with per-`ZipExtFile` locks when I found out I need to lock whole `ZipFile`.

qu1ck commented on ziprofs.py:
```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
+                    return f.read(size)
+                else:
+                    # emulate seek by reading and discarding chunks
+                    filepos = f._orig_file_size - f._left - len(f._readbuffer) + f._offset      # zipfile.py#1110 - tell()
```
Don't access hidden fields of a class you don't control.
Track the position yourself, like I did in my cache implementation.

JuniorJPDJ replied:
I'm trying to workaround this in pretty good fasion now.

qu1ck commented on ziprofs.py:
```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
+                    filepos = f._orig_file_size - f._left - len(f._readbuffer) + f._offset      # zipfile.py#1110 - tell()
+                    offset -= filepos
+                    if offset < 0:
+                        raise FuseOSError(errno.ENOTSUP)
```
I don't think you can return ENOTSUP on read calls. When you are asked to backseek on a non-seekable file reopen it and seek from start.

JuniorJPDJ replied:
If file is non-seekable every next try to open will fail, only non-seekable example of ZipExtFile is when underlying ZipFile is unseekable.
I could replace this with invalid argument error. What do you think?

JuniorJPDJ replied:
https://manpage.me/index.cgi?apropos=0&q=read&sektion=2&manpath=Debian+8.1.0&arch=default&format=html#end
You are right, I can't raise ENOSUP.

qu1ck commented on ziprofs.py:
```diff
@@ -161,24 +131,39 @@ def getattr(self, path, fh=None):
+                    if offset < 0:
+                        raise FuseOSError(errno.ENOTSUP)
+                    while offset > 0:
+                        offset -= len(f.read(min(self.MAX_SEEK_READ, offset)))
```
This loop will never end if you reach eof before offset is 0. Use my implementation for reference.

qu1ck commented on ziprofs.py:
```diff
@@ -198,37 +183,46 @@ def readdir(self, path, fh):
+            with self._zip_lock_fh[fh]:
+                ret = f.close()
+                del self._zip_file_fh[fh]
+                del self._zip_lock_fh[fh]
```
Guard this by a lock as above to make sure both entries are deleted atomically.

JuniorJPDJ replied:
This is outdated, I deleted this line.

qu1ck commented on ziprofs.py:
```diff
@@ -198,37 +183,46 @@ def readdir(self, path, fh):
-    args = parser.parse_args()
-    opts = args.opts.split(',') if args.opts else []
-    print(opts)
+    parser.add_argument('--cachesize', help="zip files cache size", type=int, default=1000)
```
This option should be part of '-o' flag, that's the only way `mount` knows how to pass options.
For example `-o cachesize=1000`.

qu1ck reviewed (changes requested):
Thanks for contributing. Looks good in general, I left comments inline.

And don't use ZipFile's internal fields, including locks. It's never a good idea to rely on implementation details of a library that can change any time and break your code.

JuniorJPDJ committed:
Locks, caching fixes, some monkey patching and better error codes.
Also fixed infinite loop on seek emulation when EOF spotted, removed usage of zipfile's internal locks, now is_zipfile caching takes mtime in account,

JuniorJPDJ commented:
> And don't use ZipFile's internal fields, including locks. It's never a good idea to rely on implementation details of a library that can change any time and break your code.

There's no reference in ZipExtFile back to ZipFile, and I don't feel like maintaining own reference list is a good idea, but that would be necessary to write this locks at my own. You see your point, but I don't feel like second way is good neither.

ATM I did pretty awful workaround, let's see what you will say about it :D

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,109 @@
-        return self.zf.infolist()
+    def open(self, *args, **kwargs):
+        zef = super(ZipFile, self).open(*args, **kwargs)
+        zef.ziplock = lambda: self.__lock
```
I'm not sure what's going on here. Why do you have 2 accessors for `__lock`? lock and ziplock do the same thing.
This isn't monkey patching, this is normal class extension, I'm not against it. I would still rather use composition instead of inheritance, like I did in my CachedZipFile.

JuniorJPDJ replied:
If I understood you correctly (I'm not sure): one accessor is on ZipFile, one is on ZipExtFile, we need the same lock on all files using the same ZipFile.

JuniorJPDJ replied:
Anyway - It just got overwritten by new `ZipExtFile` proxy subclass to make `.tell()` working for non-seekable files.

qu1ck replied:
Ah, I see, I missed that ziplock() is on ZipExtFile.

This is no good, you can get in a situation where ziplock() will try to return lock of a ZipFile which got evicted from cache and destroyed.

You should store a map of filehandles to ZipFiles and get lock from them instead of shoehorning the lock into ZipExtFile. This way you will also keep around a reference to ZipFile thus preventing it from being garbage collected even if it's evicted from cache.

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,109 @@
-                oldkey, oldvalue = self.cache.popitem(last=False)
-                self.log.debug('Popping cache entry: %s', oldkey)
-                self._cleanup(oldvalue[1])
+        with self.__lock:
```
No need to lock here because all callsites to `_add()` are in a lock anyway.

JuniorJPDJ replied:
Just in case.. If something ever changed here.. ;)

qu1ck replied:
Let's think things through and not rely on "just in case" ;)

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,109 @@
     def __init__(self, root):
         self.root = realpath(root)
-        self.rwlock = Lock()
+        # odd file handles are files inside zip, even fhs are system-wide files - collision avoidance
```
Line too long

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,109 @@
+        while True:
+            if i not in self._zip_file_fh:
+                return i
+            i += 2
```
This is still longer than it needs to be, version I suggested is more readable.

qu1ck commented on ziprofs.py:
```diff
@@ -120,23 +113,25 @@ def get_zip_path(path: str) -> str:
         for part in parts:
             cur_path = os.path.join(cur_path, part)
-            if part[-4:] == '.zip' and zipfile.is_zipfile(cur_path):
+            if part[-4:] == '.zip' and is_zipfile(cur_path):
```
I don't think is_zipfile() is needed, just call cached version directly.

JuniorJPDJ replied:
It feels much better to me with `is_zipfile()`.
It's not needed but feels better.

qu1ck replied:
One line method used only in single place should be inlined.

qu1ck commented on ziprofs.py:
```diff
@@ -162,19 +157,44 @@ def getattr(self, path, fh=None):
-            zf = self.zip_factory.get(zip_path)
-            subpath = path[len(zip_path)+1:]
-            return zf.read(subpath, size, offset)
+        if self.get_zip_path(path):
```
It's cheaper to check that fh is odd and in _zip_file_fh

JuniorJPDJ replied:
Right.

qu1ck commented on ziprofs.py:
```diff
@@ -198,37 +218,44 @@ def readdir(self, path, fh):
-        if not zip_path:
-            return os.close(fh)
-        return 0
+        if self.get_zip_path(path):
```
ditto

qu1ck commented on ziprofs.py:
```diff
@@ -162,19 +157,44 @@ def getattr(self, path, fh=None):
+                        data = len(f.read(min(self.MAX_SEEK_READ, offset)))
+                        if not data:
+                            # reached end of file - offset is beyond file
+                            raise FuseOSError(errno.EINVAL)
```
This is not invalid, return empty buffer instead.

qu1ck commented on ziprofs.py:
```diff
@@ -162,19 +157,44 @@ def getattr(self, path, fh=None):
+                    offset -= filepos
+                    if offset < 0:
+                        # cant seek back in non-seekable file
+                        raise FuseOSError(errno.EINVAL)
```
EIO seems more appropriate but I'm not sure.

JuniorJPDJ replied:
I would leave it.
https://manpage.me/index.cgi?apropos=0&q=read&sektion=2&manpath=Debian+8.1.0&arch=default&format=html
ERRORS section
IMO It seems to be most valid

JuniorJPDJ committed:
Minor speedups and fixes

JuniorJPDJ committed:
`tell()` for non-seekable ZipExtFiles and better lock

qu1ck commented on ziprofs.py:
```diff
@@ -1,116 +1,111 @@
+from typing import Optional, Dict
+
+try:
+    # noinspection PyPackageRequirements
```
Please remove these IDE specific hints. They may be useful for you but they are just clutter for me and everyone else that doesn't use your favorite IDE.
This is especially true for these annotations because in my case (debian/ubuntu) they are the wrong way around.

qu1ck commented:
I've been reading ZipFile code and I'm certain that the way we are using it ZipExtFile will allways be seekable. It's only not seekable if you initialize ZipFile with a stream file-like object instead of a path like we do.
So we can get rid of the manual position tracking and as a result custom ZipExtFile as well.

JuniorJPDJ commented:
You are right, I was trying to tell it to you.
FS will not work with non-seekable zip files, and the only case when ZipExtFile is non-seekable is non-seekable zip file.

> This is no good, you can get in a situation where ziplock() will try to return lock of a ZipFile which got evicted from cache and destroyed.

Now reference to `ZipFile` will be hold in `ZipExtFile` so we are good I bet.


I'm removing custom `ZipExtFile` and falling to pair of dicts - for `ZipFile` and `ZipExtFile` like you suggested.
Also gonna raise error when trying reads on non-seekable file.

JuniorJPDJ committed:
Falling back from custom ZipExtFile, mount options parsing, better ZipFile locks

JuniorJPDJ commented:
Is it ok now? :D

qu1ck commented on README.md:
```diff
@@ -5,8 +5,8 @@ This is a FUSE filesystem that acts as pass through to another FS except it
 ### Dependencies
-* FUSE (fuse2 works, not tested with fuse3)
-* python3-fusepy
+* FUSE
```
Did you test fuse3?

JuniorJPDJ replied:
Yes, I'm using it on fuse3 all the time.

qu1ck commented on ziprofs.py:
```diff
@@ -198,37 +194,56 @@ def readdir(self, path, fh):
+    arg = parser.parse_args()
+
+    if 'cachesize' in arg.opts:
+        CachedZipFactory.MAX_CACHE_SIZE = int(arg.opts['cachesize'])
```
Check that int is in some reasonable range, like [1, 10M]

JuniorJPDJ replied:
I'll check if it is greater than 1, I don't feel like checking upper bound is good idea.

qu1ck commented on ziprofs.py:
```diff
@@ -162,19 +146,31 @@ def getattr(self, path, fh=None):
+
+                if self.log.isEnabledFor(logging.DEBUG):
+                    foffset = f.tell()
+                    self.log.debug(f" file offset: {foffset}, read offset: {offset}, diff: {offset-foffset}")
```
Line too long. Is this needed?

JuniorJPDJ replied:
You mean debug one?
It's not needed, I used it to.. debug :V
I'll remove this in few minutes.

qu1ck reviewed (commented):
Yep, look good now. Just 2 minor nits.

JuniorJPDJ committed:
Extended README, added cache_size check, removed long debug message

qu1ck commented:
Thanks for contributing, I added better example and merged it.
Glad someone else found my small weekend project useful.

JuniorJPDJ commented:
It's the best zip FS I used until today and I was looking at whole github and google ;)
Just linked it in some places.
